packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+import { Resource, TagType } from '../cloudformation/resource';
+import { Construct, IAspect } from '../core/construct';
+
+export interface TagAspectProps extends TagProperties {
+  // TODO docs
+  value?: string;
+}
+
+export interface TagProperties {
+  // TODO docs
+  propagate?: number;
+  include?: string[];
+  exclude?: string[];
+}
+
+export abstract class TagBaseAspect implements IAspect {
+
+  public readonly type: string = 'taggable';
+
+  public readonly key: string;
+  public readonly value?: string;
+
+  public propagate: number;
+  private readonly include: string[];
+  private readonly exclude: string[];
+  private readonly visitedBy: {[id: string]: boolean} = {};
+
+  private parent?: Construct;
+
+  constructor(key: string, props: TagAspectProps = {}) {
+    this.key = key;
+
+    this.propagate = props.propagate === undefined ? -1 : props.propagate;
+    this.value = props.value;
+    this.include = props.include || [];
+    this.exclude = props.exclude || [];
+  }
+
+  public visit(node: Construct): void {
+    if (this.visitedBy[node.uniqueId] === true) {
+      return;
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
+    if (propagate !== 0) {
+      for (const child of node.children) {
+        this.propagate = propagate;
+        child.apply(this);
+      }
+    }
+    if (node instanceof Resource && (node as Resource).tagType !== TagType.NotTaggable) {
+      const resource = node as Resource;
+      if (this.exclude.length !== 0 && this.exclude.indexOf(resource.resourceType) !== -1) {
+        return;
+      }
+      if (this.include.length !== 0 && this.include.indexOf(resource.resourceType) === -1) {
+        return;
+      }
+      if (this.hasPrecedence(node.aspects)) {
+        this.applyTag(resource);
+      }
+    }
+  }
+
+  protected abstract applyTag(resource: Resource): void;
+
+  private hasPrecedence(aspects: IAspect[]): boolean {
```

Comment:
:asterisk:, if we apply in order top-down in the tree, tags applied lower in the tree will naturally overwrite the tags applied higher in the tree, without having to write this function. Let's assume the following tree:

```
┌────────────────┐                    
│                │                    
│     Parent     │    Tag(key, value1)
│                │                    
└────────────────┘                    
         │                            
         ▼                            
┌────────────────┐                    
│                │                    
│     Child      │   Tag(key, value2) 
│                │                    
└────────────────┘                    
         │                            
         ▼                            
┌────────────────┐                    
│                │                    
│TaggableResource│                    
│                │                    
└────────────────┘                            
```

This would resolve in the following way:

* The tag with `value1` would apply first. It would try to apply to `Parent` (doesn't do anything), then it would try to apply to `Child` (doesn't do anything), then it would apply to `TaggableResource` (sets `key: value1`).
* Next, the tag with `value2` would apply. It would try to apply to `Child` (doesn't do anything), then it would apply to `TaggableResource`. The value for tag `key` would be overwritten with `value2`.

Presto, the more specific tag has taken precedence without any extra work on the aspect implementor's part.